pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/HashJoinOperator.java

```diff
@@ -148,39 +172,73 @@ private TransferableBlock buildJoinedDataBlock(TransferableBlock leftBlock)
     if (leftBlock.isErrorBlock()) {
       _upstreamErrorBlock = leftBlock;
       return _upstreamErrorBlock;
-    } else if (TransferableBlockUtils.isNoOpBlock(leftBlock) || TransferableBlockUtils.isEndOfStream(leftBlock)) {
+    }
+    if (leftBlock.isNoOpBlock() || (leftBlock.isEOSBlock() && (_joinType != JoinRelType.FULL
+        && _joinType != JoinRelType.RIGHT))) {
       return leftBlock;
     }
+    if (leftBlock.isEOSBlock()) {
+      // Return remaining non-matched rows for full join in broadcast table
```

Comment:
return remaining non-matched rows for full or right OUTER joins

Reply:
Updated the comment